### PR TITLE
WT-13202 Create prefetch section in the architecture guide

### PR DESCRIPTION
**Pull request by mm-ng**

### CODE REVIEW

File: src/docs/arch-prefetch.dox [hunk 1]
```diff
@@ -0,0 +1,106 @@
+/*! @arch_page arch-prefetch Pre-fetch
+
+@section prefetch_overview What is pre-fetching? When is pre-fetching useful?
+Pre-fetching involves bringing content from the disk into the cache before it is
+required by the application in an attempt to reduce time waiting on I/O. Database
+records need to be in the cache before WiredTiger can make changes to them,
```
Comment (etienneptl) [inline src/docs/arch-prefetch.dox:6]: We don't always need to make changes, we could also simply read them

File: src/docs/arch-prefetch.dox [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+@section prefetch_other Interaction With Other Components
+
+@subsection prefetch_other_cache_busyness Considering current cache busyness
+Cache busyness increases in proportion to the eviction workload and the rate of pages
+being read into the cache. It is important to consider how pre-fetch impacts cache
+utilization as it may leave the cache in a state that increases operational latency.
+Pre-fetch stops adding page references to the queue as soon as it detects that we are
+close to hitting the eviction clean trigger. It will also skip processing page
+references that are already on the queue until space in the cache clears up.
```
Comment (etienneptl) [inline src/docs/arch-prefetch.dox:76]: I have a (personal) suggestion here, could we reword it that way?
```suggestion
references that are being retrieved from the queue until space in the cache clears up.
```

**Review by etienneptl — APPROVED** [review]
This is an excellent write up!

File: src/docs/arch-prefetch.dox [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+@subsection prefetch_other_refs Ensuring references on the pre-fetch queue remain valid
+Page references on the pre-fetch queue may become invalid if other operations
+that modify page references are happening in parallel. Eviction is not allowed to
+operate on any page references that are on the pre-fetch queue, because it may
+split pages and free the underlying ref/s. Since eviction has no way to coordinate
```
Comment (ershov) [inline src/docs/arch-prefetch.dox:82]: Eviction operates on pages residing in memory. Prefetch wants to load pages on disk to memory.
I wonder how an overlap might occur for Eviction and Prefetch to overlap. Or is it just a precaution?

Note: I don't know it this should necessarily be a part of **this** doc, I'm just curious.

File: src/docs/arch-prefetch.dox [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+- A connection level lock \c prefetch_lock is used for the pre-fetch queue, and is
+needed when adding and removing \c WT_PREFETCH_QUEUE_ENTRY items from it.
+- Whenever we modify a \c ref structure, e.g. when adding it to the pre-fetch queue
+and setting the corresponding \c WT_REF_FLAG_PREFETCH flag, it must be in the
+\c WT_REF_LOCKED state.
+- There are scenarios which fall outside of the typical eviction flow, where pages
+can be forcibly removed from the cache if eviction has exclusive eviction access to
+that file (see \c __wt_evict_file_exclusive_on). When this occurs, we need to remove
+all page references associated with that file's tree from the pre-fetch queue because
+the system may evict internal pages which still have child pages present on the queue.
```
Comment (ershov) [inline src/docs/arch-prefetch.dox:101]: Ah, I see now: Eviction should not evict any internal pages having leafs in memory. If Prefetch is going to load a leaf page, it should instruct Eviction to leave alone all parent internal pages.
Is it where the overlap can happen?
But it was mentioned that Prefetch only operates on leaf pages.
I think, I could work it out but this moment puzzles me a little bit.
Would it be worth describing this interaction in more details in this doc?

**Review by ershov — APPROVED** [review]
Thanks Monica for documenting Prefetch!
It's a great piece of doc!

File: src/docs/arch-prefetch.dox (hunk 4 above)
Comment (mm-ng) [reply src/docs/arch-prefetch.dox:101]: Yes, I wasn't sure whether I should even include this section in the first place. The pre-fetch doc is quite long already and I was thinking of cutting this out. But yes you got it correct. Eviction and pre-fetch overlap when they operate on the same refs. Happy to discuss this more if you have more questions.

File: src/docs/arch-prefetch.dox (hunk 3 above)
Comment (mm-ng) [reply src/docs/arch-prefetch.dox:82]: It's not just a precaution, and yes your other comment is correct.

File: src/docs/arch-prefetch.dox [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+tiered tables, and special operations with the exception of verify.
+
+@subsection prefetch_algorithm_populate_queue Populating the Pre-fetch Queue
+The pre-fetch queue is a first-in-first-out (FIFO) data structure containing the
+page references of potential candidates the system has identified for pre-fetching.
+The \c __wti_btree_prefetch function is responsible for pushing candidate pages
+onto the queue. Given a \c ref, it will queue the next set of pages if they are leaf
+pages, not already in the cache, and not already on the pre-fetch queue. To prevent
+overloading the pre-fetch queue, there is a maximum number of candidate pages allowed
+in the queue (\c WT_MAX_PREFETCH_QUEUE), and a limit on the maximum number of pages
```
[the remark this answers is not in the document]
Comment (mm-ng) [reply src/docs/arch-prefetch.dox:40]: True, I'll add this in. Thanks for the suggestion!

File: src/docs/arch-index.dox [hunk 6]
```diff
@@ -301,6 +302,11 @@ The format of the data file is given by structures in \c block.h .
 
 The cloud storage source extension which manage the flushing of the data files to and reading from different cloud object stores.
 
+@subpage arch-prefetch
+
+Pre-fetch is a performance optimization that aims to reduce I/O cost by reading pages from disk
```
[the remark this answers is not in the document]
Comment (mm-ng) [reply src/docs/arch-index.dox:307]: Yes, pre-fetch is targeting latency. My bad, I thought cost and latency could be used interchangeably.

File: src/docs/arch-prefetch.dox [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+we can add to the queue during a pre-fetch check (\c WT_PREFETCH_QUEUE_PER_TRIGGER).
+Pages will not be added to the queue if either of these conditions are met.
+
+@subsection prefetch_algorithm_process_pages Pre-fetching Content Into the Cache
```
Comment (zunyiliu) [inline src/docs/arch-prefetch.dox:44]: Nitpicky: I think the subsections may need to follow the same capitalization style, either title case (capitalizing the first letter of each significant word) or sentence case (capitalizing only the first letter of the first word). There're some inconsistency, e.g. here we are using title case but "@subsection prefetch_other_cache_busyness Considering current cache busyness" is using sentence case.

File: src/docs/arch-prefetch.dox [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+@subsection prefetch_api_session Session-Level Configuration
+\c WT_CONNECTION.open_session("prefetch=(enabled=true)")
+
+The session-level configuration can be used to override the connection-level
+configuration for specific sessions if desired.
+
+@section prefetch_other Interaction With Other Components
+
+@subsection prefetch_other_cache_busyness Considering current cache busyness
```
Comment (zunyiliu) [inline src/docs/arch-prefetch.dox:70]: Same here, as mentioned in above "@subsection prefetch_algorithm_process_pages Pre-fetching Content Into the Cache".

**Review by zunyiliu — APPROVED** [review]
Great documentation, thanks Monica I did learn a lot for how we document things.

**Review by ajmorton — APPROVED** [review]
LGTM!
Some small suggestions but nothing that needs to be taken. Feel free to reject them

File: src/docs/arch-prefetch.dox [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+3. Pre-fetch worker threads will pop off pages from the pre-fetch queue and read
+pages into the cache.
+
+@subsection prefetch_algorithm_find_pages Finding Candidate Pages
+Pre-fetch looks for potential candidate pages in two places: (1) while walking the
+B-tree (\c __tree_walk_internal) and (2) when verifying a tree (\c __verify_tree).
+The \c __wt_session_prefetch_check function contains a few checks to decide if
+pre-fetching should be performed for a given \c ref. Pre-fetching does not operate
+on internal pages, and should not be performed until we have triggered more than
+one leaf page read from sequential read requests. Pre-fetching is also disabled for
```
[the remark this answers is not in the document]
Comment (mm-ng) [reply src/docs/arch-prefetch.dox:30]: Yes, that's it.

**Pull request merged**